README.md
# dependent-peptides
Tools for visualising the results of dependent peptide searches
## About
Welcome to the dependent-peptides repository. Here you will find scripts and other materials relating to the research article 'Visualisation tools for dependent peptide searches to support the exploration of *in vitro* protein modifications' by Preston and co-authors. The article is freely available via the PLoS website (https://doi.org/10.1371/journal.pone.0235263).

Up-to-date information relating to the article (new findings, details of issues) will be made available via the [news](NEWS.md) file. Where appropriate, material posted to the [article's comments section](https://journals.plos.org/plosone/article/comments?id=10.1371/journal.pone.0235263) will be reproduced in this file.
## Tools
The scripts contained in this repository are tools for exploring the results of [MaxQuant](https://www.maxquant.org/) 'dependent peptide' searches. Each script processes a set of search results (one or more allPeptides.txt files) and generates a graphic. The scripts originally appeared alongside our article as supplementary files (S1-S5 Scripts). 

* Scripts I-III are for surveying the dependent peptides of a specified protein
* Scripts IV and V are for pinpointing hypothesised mass-shifts 

All scripts were written in [R for Windows](https://www.R-project.org/) and require R package [seqinr](https://cran.r-project.org/web/packages/seqinr/index.html). They were last tested on files generated by MaxQuant version 1.6.0.1, using R version 3.6.0 and seqinr version 3.4-5. 
## Getting started
Scripts I and V are the simplest scripts, and would be good ones to start with. Both scripts require two input files: (1) a FASTA file containing a single protein sequence; and (2) an allPeptides.txt file. Scripts II-IV require additional allPeptides.txt files, which they use to perform enrichment operations.

Usage notes can be found within each script's header. More detailed information can be found in our article and its supplementary files (especially file S1 Text).
* Users can check the [news](NEWS.md) file for updates regarding issues, changes and new findings. 

Example data for getting started are available via PRIDE project [PXD013040](https://www.ebi.ac.uk/pride/archive/projects/PXD013040).


NEWS.md
# News
News relating to the research article 'Visualisation tools for dependent peptide searches to support the exploration of *in vitro* protein modifications' by Preston and co-authors (https://doi.org/10.1371/journal.pone.0235263)
## Update regarding the apparent absence of BSA Cys-34 modifications
*Posted to the [article's comments section](https://journals.plos.org/plosone/article/comments?id=10.1371/journal.pone.0235263) on 17 Oct 2021*

BACKGROUND/PROBLEM. In our article, we predicted that *N*-ethylmaleimide (NEM) would modify BSA Cys-34. We tested this prediction by using Script IV to analyse the results of a series of dependent peptide (DP) searches. Script IV has the ability to link modifications to a given experimental condition (NEM treatment): firstly, it identifies DPs that were detected in both of two analyses of a test sample; then, it excludes any DPs that were detected in either of two analyses of a control sample (untreated BSA). Surprisingly, the script did not localise any putative NEM-derived modifications to Cys-34. In this update, I address the apparent absence of the expected modifications.

METHODS. MaxQuant search results were re-analysed using variants of Script IV. Two mass-shifts were considered: +68.03 Da (the mass-difference between NESyl and carbamidomethyl) and +86.04 Da (the mass-difference between hydrolysed NESyl and carbamidomethyl). The ‘treated_constant.DPs’ tables generated by Script IV were inspected. Data for DPs with the expected combination of sequence (GLVLIAFSQYLQQCPFDEHVK), modification site (Cys) and mass-shift were extracted from the search results. Precursor-ion intensities were summed across the extracted DPs, and the sums of intensities were compared across different LC-MS/MS analyses. The results of a variable-modification search were searched manually (for details of this search, see pages 3-4 of the article and file S1 Text of the supporting information).

RESULTS/DISCUSSION. Inspection of the ‘treated_constant.DPs’ tables revealed that putative NEM-derived modifications had been localised to GLVLIAFSQYLQQCPFDEHVK in all four analyses (i.e., not only in the analyses of NEM-treated BSA, but also in the control analyses). For NEM-treated BSA (analytical replicate 1), localisation probabilities tended to identify Cys-34 as the site of modification (67% of relevant features). The apparent absence of BSA Cys-34 modifications in the earlier investigation is therefore explained as follows: these modifications were in fact present, but Script IV interpreted them as ‘background’ and excluded them.

The localisation of one of the putative modifications (+86.04 Da) to Cys-34 in NEM-treated BSA was confirmed by the variable-modification search.* Notably, the quantity of relevant DPs, as judged from summed ion intensities, was always at least two orders of magnitude lower for the control sample than for NEM-treated BSA.

A plausible explanation for the above results is that the control analyses were contaminated by a small amount of ‘carry-over’ from earlier analyses.** Another, perhaps less plausible explanation is that both samples contained bona fide background modifications whose mass-shifts were the same as those of the NEM-derived modifications.

CONCLUSIONS
1. It appears that putative carry-over/interference in the control analyses caused the NEM-derived BSA Cys-34 modifications to be handled as if they were ‘background’.
2. Script IV is very sensitive to interference or contamination. The same can be assumed to be true of Scripts II and III. This is worth bearing in mind when interpreting visualisations generated by Scripts II, III and IV.
3. The ‘treated_constant.DPs’ table generated by Script IV can be useful for diagnosing problems associated with contamination and interference.

FOOTNOTES

\* The results of this search were only ever used to corroborate DP search results. This is why the Cys-34 modification did not show up in S1 Table of the supporting information.

** We did not see the putative contamination/interference in the earlier investigation because most of our analyses were focused on the ‘obvious’ NEM-derived mass shifts (i.e., +125.05 Da and +143.06 Da).

## Update regarding dependent-peptide matching
*Posted to the [article's comments section](https://journals.plos.org/plosone/article/comments?id=10.1371/journal.pone.0235263) on 14 Feb 2021*

BACKGROUND. Scripts II, III and IV perform pairwise comparisons of dependent peptides (DPs; handled as combinations of sequence and mass shift). If two DPs share the same sequence, and their mass shifts are close enough, they are considered to match.

PROBLEM. Script III is inappropriately matching DPs with the annotations ‘Comp.:13C-12C’ and ‘Deamidation’. Other problematic pairs are ‘di-Oxidation/Sulfide’ and ‘Comp.:S-NH3O/Comp.:O-NH3’. Further pairs with very close shifts could also be problematic. Script II matches DPs using the same basic method as Script III, and will be similarly affected. Script IV uses a different method, and may not be affected to the same extent.

SOLUTIONS. For Script III, I have identified three potential solutions: (i) excluding interfering modifications (e.g., ‘Comp.:13C-12C’); (ii) requiring annotations to match; and (iii) halving the tolerance. I tested these solutions by making variants of Script III and reprocessing the BSA data reported in the paper. A match was considered inappropriate if the DPs’ annotations were different. The matching of unannotated DPs with different mass shifts was not tested for presently (however, see CONCLUSIONS).


Script_I.R
# +-----------------------------------------------------------------------------------+
# | Supporting information file S1 Script                                             |
# +-----------------------------------------------------------------------------------+
# | Script I                                                                          |
# |                                                                                   |
# | To accompany the report 'Visualisation tools for dependent peptide searches to    |
# | support the exploration of in vitro protein modifications' by G. W. Preston, L.   |
# | Yang, D. H. Phillips and C. S. Maier                                              |
# +-----------------------------------------------------------------------------------+

#   Usage notes:
#   1. This script filters the dependent peptides from a given analysis (e.g. an 
#      analysis of NEM-treated BSA) and generates a dependent-peptide localisation 
#      plot and a mass-shift frequency histogram.
#   2. The script requires:
#      i.   R version 3.6.0 or later;
#      ii.  Windows (a Windows-specific graphics device will be used);
#      iii. package 'seqinR';
#      iv.  one allPeptides.txt file;
#      v.   a *.fasta file containing the sequence of a protein of interest;
#      vi.  the ID (e.g. UniProt ID) of the protein of interest, exactly as it 
#           appears in the allPeptides.txt file.
#   3. Data from our BSA study, including the allPeptides.txt and *.fasta files 
#      specified in the script, are available via PRIDE/ProteomeXchange (accession 
#      PXD013040).
#   4. Further instructions are provided within the script. Comments beginning with a
#      greater-than symbol (>) are instructions for users.
#   5. Detailed explanatory notes can be found in Tables B and C of supporting 
#      information file S1 Text.

# +-----------------------------------------------------------------------------------+

# Choose options
options("stringsAsFactors" = FALSE)

# Load seqinR
library(seqinr)

# Get protein sequence
# > Specify the path to a *.fasta file
protein.sequence <- read.fasta(paste("C:", "\\", "Users", "\\", "George", "\\",
	"MaxQuant_1.6.0.1", "\\", "Databases", "\\", "4f5s_A.fasta", sep = ""),
	seqtype = "AA",
	seqonly = TRUE)[[1]]

# > Specify protein of interest
protein.ID <- "4F5S"

# Get search results
# > Specify the path to an allPeptides.txt file
all.peptides <- read.table(paste("C:", "\\", "Users", "\\", "George", "\\",
	"Documents", "\\", "Model data", "\\", "treated_replicate1_allPeptides.txt", 
	sep = ""), 
	sep = "\t", 
	header = TRUE,
	fill = TRUE)

# Make a non-redundant list of observed peptide sequences
peptide.sequences <- unique(c(all.peptides$Sequence[grepl(protein.ID, 
	all.peptides$Proteins) &
	all.peptides$Score > 80],
	all.peptides$DP.Base.Sequence[
		all.peptides$DP.Score > 80 &
		all.peptides$DP.PEP < 0.01 &
		all.peptides$DP.Decoy != "+" &
		grepl(protein.ID, all.peptides$DP.Proteins)]))

# Filter search results
dependent.peptides <- all.peptides[!is.na(all.peptides$DP.Mass.Difference) &
	all.peptides$DP.Mass.Difference < 500.5 &
	all.peptides$DP.Mass.Difference >= -500.5 &
	all.peptides$DP.Score > 80 &	
	all.peptides$DP.PEP < 0.01 &
	all.peptides$DP.Decoy != "+" &
	grepl(protein.ID, all.peptides$DP.Proteins) &
	!(all.peptides$DP.Modification %in% c("Carbamidomethyl",
		"Loss of ammonia",
		"Loss of water",
		"Unmodified")) &
	!(all.peptides$DP.Modification == "CarbamidomethylDTT" &
	all.peptides$DP.AA == "C") &
	!grepl("Cation:", all.peptides$DP.Modification) &
	all.peptides$DP.Peptide.Length.Difference >= 0, ]

# Choose a graphics device and specify the plotting-area dimensions
windows(7, 3)

# Define custom plot margins, layout and common axis limits
par(mar = c(5, 4, 1, 2) + 0.1)
layout(matrix(c(1, 1, 1, 1, 2, 2, 2), 
	nrow = 1, 
	ncol = 7), 
	respect = FALSE)
delta.mass.axis.limits <- c(-500, 500)

# Create an empty left-hand plot
plot(1,	0, 
	xlim = c(0, nchar(protein.sequence)),	
	ylim = delta.mass.axis.limits,
	type = "n",
	xlab = "Amino acid residue",
	ylab = expression(paste(Delta, "m / Da", sep = "")),
	las = 1)

# Add a dashed 'backbone'
lines(c(0.5, nchar(protein.sequence) + 0.5), 
	c(0, 0),
	lty = 3, 
	col = "red")

# Solidify observed regions of the backbone
for(i in 1:length(peptide.sequences))
	{for(j in 1:(nchar(protein.sequence) -
	(nchar(peptide.sequences[i]) - 1)))
	{segment <- substr(protein.sequence, 
		start = j, 
		stop = j + (nchar(peptide.sequences[i]) - 1))
	if(segment == peptide.sequences[i])
	lines(c(j - 0.5,
		j + (nchar(peptide.sequences[i]) - 1) + 0.5),
		c(0, 0), 
		col = "red")}}

# Add dependent peptides
for(i in 1:nrow(dependent.peptides))
	{for(j in 1:(nchar(protein.sequence) - 
	(nchar(dependent.peptides$DP.Base.Sequence[i]) - 1)))
	{segment <- substr(protein.sequence,
		start = j, 
		stop = j + (nchar(dependent.peptides$DP.Base.Sequence[i]) - 1))
	if(segment == dependent.peptides$DP.Base.Sequence[i])
	{vertices <- matrix(data = c(j - 0.5,
		j + (nchar(dependent.peptides$DP.Base.Sequence[i]) - 1) + 0.5,
		j + (nchar(dependent.peptides$DP.Base.Sequence[i]) - 1) + 0.5,
		j - 0.5,
		0,
		0, 
		dependent.peptides$DP.Mass.Difference[i], 
		dependent.peptides$DP.Mass.Difference[i]),
		nrow = 4,
		ncol = 2)
	polygon(vertices, 
		border = gray(level = 0, alpha = 0.5))}}}

# Prepare histogram
histogram <- hist(dependent.peptides$DP.Mass.Difference,
	breaks = seq(-500.5, 500.5, 1),
	right = FALSE,
	plot = FALSE)

# Draw histogram
plot(histogram, 
	main = NULL,
	xlim = delta.mass.axis.limits,
	ylim = c(0, max(histogram$counts) * 1.05),
	xlab = expression(paste(Delta, "m / Da", sep = "")),
	ylab = "Number of peptides",
	las = 1)
box()

# Specify how the histogram should be annotated
# > You could vary the annotation count
# > Type 'histogram.slices' to view the table
histogram.slices <- data.frame(Slice.number = 1:13, 
	First.cell = c(1, 101, 201, 301, 401, 451, 501, 
		502, 552, 602, 702, 802, 902),
	Last.cell = c(100, 200, 300, 400, 450, 500, 501, 
		551, 601, 701, 801, 901, 1001),
	Mass.difference.range = numeric(13),
	Annotation.count = c(0, 0, 0, 1, 3, 3, 1, 3, 3, 1, 1, 0, 0))
for(i in 1:nrow(histogram.slices))
histogram.slices$Mass.difference.range[i] <- paste(
	as.character(histogram$mids[histogram.slices$First.cell[i]]),
	"to",
	as.character(histogram$mids[histogram.slices$Last.cell[i]]))

# Annotate the histogram
for(i in 1:nrow(histogram.slices))
{slice_counts <- histogram$counts[histogram.slices$First.cell[i]:
	histogram.slices$Last.cell[i]]
if(sum(slice_counts) > 0 & histogram.slices$Annotation.count[i] > 0)
{cells.to.annotate <- histogram.slices$First.cell[i] + 
	order(-slice_counts)[1:histogram.slices$Annotation.count[i]] - 1
text(histogram$mids[cells.to.annotate], 
	histogram$counts[cells.to.annotate] + max(histogram$counts) / 19, 
	labels = as.character(histogram$mids[cells.to.annotate]))}}

# +-----------------------------------------------------------------------------------+

Script_IV.R
# +-----------------------------------------------------------------------------------+
# | Supporting information file S4 Script                                             |
# +-----------------------------------------------------------------------------------+
# | Script IV                                                                         |
# |                                                                                   |
# | To accompany the report 'Visualisation tools for dependent peptide searches to    |
# | support the exploration of in vitro protein modifications' by G. W. Preston, L.   |
# | Yang, D. H. Phillips and C. S. Maier                                              |
# +-----------------------------------------------------------------------------------+

#   Usage notes:
#   1. This script isolates dependent peptides with a specified mass shift, enriches 
#      them on the basis of 'constant conjunction' (see note 2) and generates a 
#      probability localisation plot.
#   2. In order to be visualised, a dependent peptide must be 'constantly conjoined'
#      with the experimental condition (e.g. NEM treatment); that is, it must:
#      - have been detected in the given analysis;
#      - have been detected in a second analysis of the same sample; 
#      - have gone undetected in both of two analyses of a second sample (e.g. a 
#        sample of untreated BSA).
#   3. The script requires:
#      i.   R version 3.6.0 or later;
#      ii.  Windows (a Windows-specific graphics device will be used);
#      iii. package 'seqinR';
#      iv.  four allPeptides.txt files (e.g. two for NEM-treated BSA and two for
#           untreated BSA);
#      v.   a *.fasta file containing the sequence of a protein of interest;
#      vi.  the ID (e.g. UniProt ID) of the protein of interest, exactly as it 
#           appears in the allPeptides.txt files;
#      vii. an expected mass shift and a tolerance. We use a tolerance of 0.01 Da.
#   4. Data from our BSA study, including the allPeptides.txt and *.fasta files 
#      specified in the script, are available via PRIDE/ProteomeXchange 
#      (accession PXD013040).
#   5. Further instructions are provided within the script. Comments beginning with a
#      greater-than symbol (>) are instructions for users.
#   6. Detailed explanatory notes can be found in Tables B and C of supporting 
#      information file S1 Text. 

# +-----------------------------------------------------------------------------------+

# Choose options
options("stringsAsFactors" = FALSE)

# Load seqinR
library(seqinr)

# Get protein sequence
# > Specify the path to a *.fasta file
protein.sequence <- read.fasta(paste("C:", "\\", "Users", "\\", "George", "\\",
	"MaxQuant_1.6.0.1", "\\", "Databases", "\\", "4f5s_A.fasta", sep = ""),
	seqtype = "AA",
	seqonly = TRUE)[[1]]

# > Specify protein of interest
protein.ID <- "4F5S"

# > Input expected mass shift
expected.delta.mass <- 143.05825

# > Specify tolerance
tolerance <- 0.01

# Get search results (treated protein, analysis 1)
# > Specify the path to an allPeptides.txt file
treated_replicate1_all.peptides <- read.table(paste("C:", "\\", "Users", "\\", 
	"George", "\\",	"Documents", "\\", "Model data", "\\", 
	"treated_replicate1_allPeptides.txt", sep = ""), 
	sep = "\t", 
	header = TRUE,
	fill = TRUE)

# Get search results (treated protein, analysis 2)
# > Specify the path to an allPeptides.txt file
treated_replicate2_all.peptides <- read.table(paste("C:", "\\", "Users", "\\", 
	"George", "\\", "Documents", "\\", "Model data", "\\", 
	"treated_replicate2_allPeptides.txt", sep = ""), 
	sep = "\t", 
	header = TRUE,
	fill = TRUE)

# Get search results (untreated protein, analysis 1)
# > Specify the path to an allPeptides.txt file
control_replicate1_all.peptides <- read.table(paste("C:", "\\", "Users", "\\", 
	"George", "\\", "Documents", "\\", "Model data", "\\", 
	"control_replicate1_allPeptides.txt", sep = ""), 
	sep = "\t", 
	header = TRUE,
	fill = TRUE)

# Get search results (untreated protein, analysis 2)
# > Specify the path to an allPeptides.txt file
control_replicate2_all.peptides <- read.table(paste("C:", "\\", "Users", "\\", 
	"George", "\\", "Documents", "\\", "Model data", "\\", 
	"control_replicate2_allPeptides.txt", sep = ""), 
	sep = "\t", 
	header = TRUE,
	fill = TRUE)

# Make a non-redundant list of peptide sequences (untreated protein, analysis 1)
control_replicate1_peptide.sequences <- unique(c(
	control_replicate1_all.peptides$Sequence[grepl(protein.ID, 
		control_replicate1_all.peptides$Proteins) &
		control_replicate1_all.peptides$Score > 80],
	control_replicate1_all.peptides$DP.Base.Sequence[grepl(protein.ID, 
		control_replicate1_all.peptides$DP.Proteins) &
		control_replicate1_all.peptides$DP.Score > 80 &
		control_replicate1_all.peptides$DP.PEP < 0.01 &
		control_replicate1_all.peptides$DP.Decoy != "+"]))

# Make a non-redundant list of peptide sequences (untreated protein, analysis 2)
control_replicate2_peptide.sequences <- unique(c(
	control_replicate2_all.peptides$Sequence[grepl(protein.ID, 
		control_replicate2_all.peptides$Proteins) &
		control_replicate2_all.peptides$Score > 80],
	control_replicate2_all.peptides$DP.Base.Sequence[grepl(protein.ID, 
		control_replicate2_all.peptides$DP.Proteins) &
		control_replicate2_all.peptides$DP.Score > 80 &
		control_replicate2_all.peptides$DP.PEP < 0.01 &
		control_replicate2_all.peptides$DP.Decoy != "+"]))

# Compile sequences that are common to both lists
control_common.peptide.sequences <- control_replicate1_peptide.sequences[
	control_replicate1_peptide.sequences %in% 
		control_replicate2_peptide.sequences]

# Filter search results (treated protein, analysis 1)
treated_replicate1_dependent.peptides <- treated_replicate1_all.peptides[
	!is.na(treated_replicate1_all.peptides$DP.Mass.Difference) &
	treated_replicate1_all.peptides$DP.Mass.Difference < 
		expected.delta.mass + tolerance &
	treated_replicate1_all.peptides$DP.Mass.Difference > 
		expected.delta.mass - tolerance &
	treated_replicate1_all.peptides$DP.Score > 80 &	
	treated_replicate1_all.peptides$DP.PEP < 0.01 &
	treated_replicate1_all.peptides$DP.Decoy != "+" &
	grepl(protein.ID, treated_replicate1_all.peptides$DP.Proteins), ]

# Filter search results (treated protein, analysis 2)
treated_replicate2_dependent.peptides <- treated_replicate2_all.peptides[
	!is.na(treated_replicate2_all.peptides$DP.Mass.Difference) &
	treated_replicate2_all.peptides$DP.Mass.Difference < 
		expected.delta.mass + tolerance &
	treated_replicate2_all.peptides$DP.Mass.Difference > 
		expected.delta.mass - tolerance &
	treated_replicate2_all.peptides$DP.Score > 80 &	
	treated_replicate2_all.peptides$DP.PEP < 0.01 &
	treated_replicate2_all.peptides$DP.Decoy != "+" &
	grepl(protein.ID, treated_replicate2_all.peptides$DP.Proteins), ]

# Do not proceed any further if either of the last two data frames are empty
if(nrow(treated_replicate1_dependent.peptides) > 0 &
	nrow(treated_replicate2_dependent.peptides) > 0)

# Filter search results (untreated protein, analysis 1)
{control_replicate1_dependent.peptides <- control_replicate1_all.peptides[
	!is.na(control_replicate1_all.peptides$DP.Mass.Difference) &
	control_replicate1_all.peptides$DP.Mass.Difference < 
		expected.delta.mass + tolerance &
	control_replicate1_all.peptides$DP.Mass.Difference > 
		expected.delta.mass - tolerance &
	control_replicate1_all.peptides$DP.Score > 80 &	
	control_replicate1_all.peptides$DP.PEP < 0.01 &
	control_replicate1_all.peptides$DP.Decoy != "+" &
	grepl(protein.ID, control_replicate1_all.peptides$DP.Proteins), ]

# Filter search results (untreated protein, analysis 2)
control_replicate2_dependent.peptides <- control_replicate2_all.peptides[
	!is.na(control_replicate2_all.peptides$DP.Mass.Difference) &
	control_replicate2_all.peptides$DP.Mass.Difference < 
		expected.delta.mass + tolerance & 
	control_replicate2_all.peptides$DP.Mass.Difference > 
		expected.delta.mass - tolerance &
	control_replicate2_all.peptides$DP.Score > 80 &	
	control_replicate2_all.peptides$DP.PEP < 0.01 &
	control_replicate2_all.peptides$DP.Decoy != "+" &
	grepl(protein.ID, control_replicate2_all.peptides$DP.Proteins), ]

# Make empty data frame for sequences, 'DP Probabilities' strings and test outcomes
treated_DP.constancy.results <- data.frame(
	DP.Base.Sequence = character(nrow(treated_replicate1_dependent.peptides)),
	DP.Probabilities.1 = character(nrow(treated_replicate1_dependent.peptides)),
	DP.Probabilities.2 = character(nrow(treated_replicate1_dependent.peptides)),
	In.control_replicate1 = character(nrow(treated_replicate1_dependent.peptides)),
	In.control_replicate2 = character(nrow(treated_replicate1_dependent.peptides)))
for(i in 1:nrow(treated_DP.constancy.results))
	{for(j in 1:ncol(treated_DP.constancy.results))
	treated_DP.constancy.results[i, j] <- NA}

# Perform first round of pairwise comparisons (treated <-> treated), deposit data
for(i in 1:nrow(treated_replicate1_dependent.peptides))
	{for(j in 1:nrow(treated_replicate2_dependent.peptides))
	if(treated_replicate1_dependent.peptides$DP.Base.Sequence[i] ==
		treated_replicate2_dependent.peptides$DP.Base.Sequence[j])
	{treated_DP.constancy.results$DP.Base.Sequence[i] <- 
		treated_replicate1_dependent.peptides$DP.Base.Sequence[i]
	treated_DP.constancy.results$DP.Probabilities.1[i] <- 
		treated_replicate1_dependent.peptides$DP.Probabilities[i]
	treated_DP.constancy.results$DP.Probabilities.2[i] <- 
		treated_replicate2_dependent.peptides$DP.Probabilities[j]}}

# Remove rows of NAs
treated_constant.DPs <- treated_DP.constancy.results[
	!is.na(treated_DP.constancy.results$DP.Base.Sequence), ]

# Perform second round of comparisons (treated <-> untreated), deposit outcomes
if(nrow(control_replicate1_dependent.peptides) > 0)
for(i in 1:nrow(treated_constant.DPs))
	{for(j in 1:nrow(control_replicate1_dependent.peptides))
	if(!is.na(control_replicate1_dependent.peptides$DP.Base.Sequence[j]) &
		treated_constant.DPs$DP.Base.Sequence[i] ==
		control_replicate1_dependent.peptides$DP.Base.Sequence[j])
	treated_constant.DPs$In.control_replicate1[i] <- "Yes"}

# Perform third round of comparisons (treated <-> untreated), deposit outcomes
if(nrow(control_replicate2_dependent.peptides) > 0)
for(i in 1:nrow(treated_constant.DPs))
	{for(j in 1:nrow(control_replicate2_dependent.peptides))
	if(!is.na(control_replicate2_dependent.peptides$DP.Base.Sequence[j]) &
		treated_constant.DPs$DP.Base.Sequence[i] ==
		control_replicate2_dependent.peptides$DP.Base.Sequence[j])
	treated_constant.DPs$In.control_replicate2[i] <- "Yes"}

# Add negative outcomes
for(i in 1:nrow(treated_constant.DPs))
	{for(j in 4:5)
	if(is.na(treated_constant.DPs[i, j]))
	treated_constant.DPs[i, j] <- "No"}

# Isolate 'constantly conjoined' dependent peptides
treated_constantly.conjoined.DPs <- treated_constant.DPs[
	treated_constant.DPs$In.control_replicate1 == "No" &
	treated_constant.DPs$In.control_replicate2 == "No", ]

# Prepare empty matrices
transparency <- matrix(nrow = nchar(protein.sequence), ncol = 11)
reservations <- matrix(nrow = nchar(protein.sequence), ncol = 11)
protein.probabilities <- matrix(0, nrow = nchar(protein.sequence), ncol = 11)

# Write data into matrices via sliding window
for(i in 1:nrow(treated_constantly.conjoined.DPs))
	{for(j in 1:(nchar(protein.sequence) - 
	(nchar(treated_constantly.conjoined.DPs$DP.Base.Sequence[i]) - 1)))
	{segment <- substr(protein.sequence, 
		start = j, 
		stop = j + (nchar(
			treated_constantly.conjoined.DPs$DP.Base.Sequence[i]) - 1))

	if(segment == treated_constantly.conjoined.DPs$DP.Base.Sequence[i])
	
	# Define strips
	{footprint <- j:(j + (nchar(
		treated_constantly.conjoined.DPs$DP.Base.Sequence[i]) - 1))
	oversized.footprint <- ((j - 1):
		(j + length(footprint)))[((j - 1):(j + length(footprint))) %in%
			1:nchar(protein.sequence)]
	
	# Parse 'DP Probabilities'
	primary.fragments <- strsplit(
		treated_constantly.conjoined.DPs$DP.Probabilities.1[i],
		split = "(", 
		fixed = TRUE)[[1]]
	peptide.probabilities <- data.frame(
		Sequence = character(length(primary.fragments)), 
		Probability = numeric(length(primary.fragments)),
		Peptide.site = numeric(length(primary.fragments)),
		Protein.site = numeric(length(primary.fragments)))
	peptide.probabilities$Sequence[1] <- primary.fragments[1]
	peptide.probabilities$Peptide.site[1] <- nchar(primary.fragments[1])
		for(k in 2:length(primary.fragments))
		{secondary.fragments <- strsplit(primary.fragments[k], 
			split = ")", 
			fixed = TRUE)[[1]]
		peptide.probabilities$Probability[k - 1] <- as.numeric(
			secondary.fragments[1])
		peptide.probabilities$Sequence[k] <- secondary.fragments[2]
		peptide.probabilities$Peptide.site[k] <- 
			peptide.probabilities$Peptide.site[k - 1] + 
				nchar(peptide.probabilities$Sequence[k])
		peptide.probabilities$Protein.site <- 
			peptide.probabilities$Peptide.site + j - 1}
	
	# Allocate an unreserved strip, write data into matrices
	if(length(which(is.na(reservations[oversized.footprint, 5]))) == 
		length(oversized.footprint))
		{for(k in 1:nrow(peptide.probabilities))
			{for(m in 1:nrow(protein.probabilities))
			if(!is.na(peptide.probabilities$Protein.site[k]) & 
				peptide.probabilities$Protein.site[k] == m)
			protein.probabilities[m, 5] <- 
				peptide.probabilities$Probability[k]}
			reservations[oversized.footprint, 5] <- 1
			transparency[footprint, 5] <- 1} else
	if(length(which(is.na(reservations[oversized.footprint, 4]))) == 
		length(oversized.footprint))
		{for(k in 1:nrow(peptide.probabilities))
			{for(m in 1:nrow(protein.probabilities))
			if(!is.na(peptide.probabilities$Protein.site[k]) & 
				peptide.probabilities$Protein.site[k] == m)
			protein.probabilities[m, 4] <- 
				peptide.probabilities$Probability[k]}
			reservations[oversized.footprint, 4] <- 1
			transparency[footprint, 4] <- 1} else
	if(length(which(is.na(reservations[oversized.footprint, 3]))) == 
		length(oversized.footprint))
		{for(k in 1:nrow(peptide.probabilities))
			{for(m in 1:nrow(protein.probabilities))
			if(!is.na(peptide.probabilities$Protein.site[k]) & 
				peptide.probabilities$Protein.site[k] == m)
			protein.probabilities[m, 3] <- 
				peptide.probabilities$Probability[k]}
			reservations[oversized.footprint, 3] <- 1
			transparency[footprint, 3] <- 1} else
	if(length(which(is.na(reservations[oversized.footprint, 2]))) == 
		length(oversized.footprint))
		{for(k in 1:nrow(peptide.probabilities))
			{for(m in 1:nrow(protein.probabilities))
			if(!is.na(peptide.probabilities$Protein.site[k]) & 
				peptide.probabilities$Protein.site[k] == m)
			protein.probabilities[m, 2] <- 
				peptide.probabilities$Probability[k]}
			reservations[oversized.footprint, 2] <- 1
			transparency[footprint, 2] <- 1} else
	if(length(which(is.na(reservations[oversized.footprint, 1]))) == 
		length(oversized.footprint))
		{for(k in 1:nrow(peptide.probabilities))
			{for(m in 1:nrow(protein.probabilities))
			if(!is.na(peptide.probabilities$Protein.site[k]) & 
				peptide.probabilities$Protein.site[k] == m)
			protein.probabilities[m, 1] <- 
				peptide.probabilities$Probability[k]}
			reservations[oversized.footprint, 1] <- 1
			transparency[footprint, 1] <- 1}}}}

# Calculate required number of panels 
panel.count <- ceiling(nchar(protein.sequence) / 100)

# Choose a graphics device and specify plotting-area dimensions
windows(14, 10)

# Define layout
if(panel.count < 6)
par(mfcol = c(6, 1)) else 
par(mfcol = c(panel.count, 1))

# Specify custom plot margins
par(mar = c(3, 5, 0, 5))

# Prepare the graphic, one panel at a time
for(i in 1:panel.count)

# Make composite image
{image(x = 1:nchar(protein.sequence), 
	z = protein.probabilities,
	breaks = seq(0, 1, 0.05), 
	col = gray(seq(1, 0.05, -0.05)),
	xlim = c(i * 100 - 99.5, 
	i * 100 + 0.5),
	axes = FALSE,
	ann = FALSE)
image(x = 1:nchar(protein.sequence), 
	z = transparency,
	breaks = seq(0, 1, 1), 
	col = rainbow(1, start = 2/3, alpha = 0.5), add = TRUE,
	xlim = c(i * 100 - 99.5, 
	i * 100 + 0.5),
	axes = FALSE,
	ann = FALSE)
axis(side = 1)

# Add a dashed 'backbone'
lines(c(0.5, nchar(protein.sequence) + 0.5), 
	c(0.5, 0.5),
	lty = 3)

	# Solidify observed regions of the backbone
	for(j in 1:length(control_common.peptide.sequences))
		{for(k in 1:(nchar(protein.sequence) -
		(nchar(control_common.peptide.sequences[j]) - 1)))
		{segment <- substr(protein.sequence, 
			start = k, 
			stop = k + (nchar(control_common.peptide.sequences[j]) - 1))
		if(segment == control_common.peptide.sequences[j])
		lines(c(k - 0.5, k + (nchar(
			control_common.peptide.sequences[j]) - 1) + 0.5),
				c(0.5, 0.5),
				lty = 1,
				col = "red")}}
	
	# Add annotations
	for(j in 1:nrow(protein.probabilities))
	if(sum(protein.probabilities[j, ]) > 0)
	text(j, 0.8, labels = paste(substr(protein.sequence, start = j, stop = j),
		as.character(j),
		sep = ""),
		srt = 90)

box()}} else
print("Expected delta mass not observed in at least one analysis of treated protein")

# +-----------------------------------------------------------------------------------+